_drafts/restic-foundation1-cdc.md

```diff
@@ -0,0 +1,268 @@
+backup, this is where the term "incremental" backup comes from. This way,
+unmodified files are not stored again on subsequent backups. But what happens if
+just a small portion of a large file is modified? Using this strategy, the
+modified file will be saved again, although most of it did not change.
+
+A better idea is splitting files into smaller fixed-size pieces (called
+"chunks" in the following) of e.g. 1MiB in size. When the backup program saves a
+file to the backup location, it is sufficient to save all chunks and the list of
+chunks the file consists of. These chunks can be identified for example by the
+SHA-256 hash of the content, so duplicate chunks can be detected and saved only
+once. This way, for a file consisting of a large number of null bytes, only one
+chunk of null bytes needs to be stored.
+
+On a subsequent backup, unmodified files are not saved again because all chunks
+have already been saved before. Modified files on the other hand are split into
+chunks again, and new chunks are saved to the backup location.
+
+But what happens when the user adds a byte to the beginning of the file? The
+chunk-boundaries would shift by one byte, changing every chunk in the file.
+When the backup program now splits the file into fixed-sized chunks, it would
+(in most cases) end up with a list of different chunks, so it needs to save
+every chunk as a new chunk to the backup location. This is not satisfactory for
+a modern backup program.
+
+### Content Defined Chunking
+
+Restic works a bit differently. It also operates on chunks of data from files
+and only upload new chunks, but uses a more sophisticated approach for
+splitting files into chunks called *Content Defined Chunking*. It works by
+splitting a file into chunks based on the contents of the file, rather than
+always splitting after a fixed number of bytes.
+
+In the following, the function $$F(b_0 \dots b_{63})$$ returns a 64 bit
+[Rabin Fingerprint](https://en.wikipedia.org/wiki/Rabin_fingerprint)
+of the byte sequence in the argument. This function can be efficiently computed
+as a [rolling hash](https://en.wikipedia.org/wiki/Rolling_hash), which means
+that $$F(b_1\dots b_{64})$$ can be computed without much overhead when
+$$F(b_0\dots b_{63})$$ is already known. Restic uses 64 bytes as the "window
+size" for the rolling hash.
+
+When restic saves a file, it first computes the Rabin Fingerprints for all 64
+byte sequences in the file, so it starts by computing $$F(b_0\dots b_{63})$$,
+then $$F(b_1\dots b_{64})$$, then $$F(b_2\dots b_{65})$$ and so on. For each
+fingerprint, restic then tests if the lowest 21 bits are zero. If this is the
+case, restic found a new chunk boundary.
+
+A chunk boundary therefore depends only on the last 64 bytes before the
+boundary, in other words the end of a chunk depends on the last 64 bytes of a
+chunk. Returning to our earlier example, if a user creates a backup of a file
```

Comment:
> chunk boundary therefore depends only on the last 64 bytes

Maybe make it more explicit that this means that chunks can have different sizes depending on the content?